### Урок 5. Проверка статистических гипотез. P-значения. Доверительные интервалы. A/B-тестирование

In [1]:
import numpy as np
from scipy import stats

##### Задача 1

Известно, что генеральная совокупность распределена нормально с известным средним квадратическим отклонением 16. Найти доверительный интервал для оценки математического ожидания mu с надёжностью 0.95, если выборочное среднее равно 80, а объём равен 256.

Поскольку рассматриваемая случайная величина имеет нормальное распределение и известно среднее квадратичное отклонение (соответственно и дисперсия тоже известна), то доверительный интервал для математического ожидания можно построить используя статистику Z и квантили стандартного распределения. 

$$Z = \dfrac{\overline{X} - \mu}{\sigma / \sqrt{n}},$$

где $\sigma$ — известный параметр среднего квадратического отклонения. Критическая область будет двухсторонней. Для её построения нам понадобятся квантили  𝑡𝛼/2  и  𝑡1−𝛼/2  стандартного нормального распределения.

$$P \left( t_{\alpha / 2} \leq Z \leq t_{1 - \alpha / 2} \right) = p,$$

Подставим сюда $Z$ и получим:

$$P \left( t_{\alpha / 2} \leq \dfrac{\overline{X} - \mu}{\sigma / \sqrt{n}} \leq t_{1 - \alpha / 2} \right) = p$$
$$P \left( t_{\alpha / 2} \cdot \dfrac{\sigma}{\sqrt{n}} \leq \overline{X} - \mu \leq t_{1 - \alpha / 2} \cdot \dfrac{\sigma}{\sqrt{n}} \right) = p$$
<center>
    <div style="width: 350px;border:1px solid black;">
        $P \left( \overline{X} + t_{\alpha / 2} \cdot \dfrac{\sigma}{\sqrt{n}} \leq \mu \leq \overline{X} + t_{1 - \alpha / 2} \cdot \dfrac{\sigma}{\sqrt{n}} \right) = p$
    </div>
</center>

In [31]:
mean = 80
n = 256
p = 0.95
sigma = 16
alpha = 1 - p

In [32]:
t1 = stats.norm.ppf(alpha / 2)
t2 = stats.norm.ppf(1 - alpha / 2)

t1, t2

(-1.959963984540054, 1.959963984540054)

И тогда доверительный интервал:

In [33]:
(mean + t1 * sigma / np.sqrt(n), mean + t2 * sigma / np.sqrt(n))

(78.04003601545995, 81.95996398454005)

#### Задача 2

Продавец утверждает, что средний вес пачки печенья составляет 200 г. Из партии извлечена выборка из 10 пачек. Вес каждой пачки составляет:

202, 203, 199, 197, 195, 201, 200, 204, 194, 190
Известно, что их веса распределены нормально.
1. Верно ли утверждение продавца, если учитывать, что уровень значимости равен 1%?
2. Найдите P-значение для данного теста.

In [12]:
weights = np.array([202, 203, 199, 197, 195, 201, 200, 204, 194, 190])
n = 10

mean = weights.mean()
std = weights.std(ddof=1)

mean, std

(198.5, 4.453463071962462)

Поскольку мы проверяем гипотезу о математическом ожидании нормально распределённой случайной величины с неизвестной дисперсией, то будем использовать $T$статистику:

$$t = \dfrac{\overline{X} - \mu}{\sigma_X / \sqrt{n}},$$

Рассчитаем $T$статистику:

In [13]:
T = (mean-200)/(std/np.sqrt(n))
T

-1.0651074037450896

Критическая область:

In [14]:
alpha = 0.01

t1 = stats.t.ppf(alpha / 2, df=n - 1)
t2 = stats.t.ppf(1 - alpha / 2, df=n - 1)

t1, t2

(-3.2498355440153697, 3.2498355440153697)

Итак, теперь критическая область:
$$\Omega_\alpha = (-\infty, -3.24) \cup (3.24, \infty)$$

Поскольку значение $T$ не попадает в критическую область, то утверждение продавца верно.

In [17]:
p_left = stats.t.cdf(T, df=n - 1)
p_right = 1 - stats.t.cdf(T, df=n - 1)

pvalue = 2 * min(p_left, p_right)

pvalue

0.3145666035247285

Поскольку pvalue больше, чем $\alpha = 0.01$, то можно сделать вывод, что утверждение продавца так же верно с вероятностью 99% 

#### Задача 3

Утверждается, что шарики для подшипников, изготовленные автоматическим станком, имеют средний диаметр 17 мм. Используя уровень значимости 0.025 и правостороннюю критическую область, проверить данную гипотезу при условии, что в выборке из 100 шариков средний диаметр оказался равным 17.5 мм, а дисперсия известна и равна 4 мм.

In [20]:
n = 100
d = 2
samples = np.random.normal(17.5, d, size=n)

samples

array([18.46578637, 16.67206004, 19.83282478, 16.92307731, 18.52646117,
       16.2812967 , 20.18717668, 16.24059447, 19.51363773, 17.64025933,
       17.76259673, 18.90824255, 18.29397653, 17.07680207, 19.38953484,
       18.22926953, 16.53122911, 17.3104733 , 16.43384754, 16.03726791,
       17.20017358, 16.99084855, 12.66973259, 18.66314387, 19.35806286,
       17.91621215, 18.75686204, 18.71146458, 16.60814143, 20.62131284,
       17.24293659, 16.47650619, 19.59769318, 19.21670298, 18.55730267,
       20.83130465, 19.31096354, 20.33442093, 18.31457223, 16.05249122,
       18.60028638, 17.56286695, 16.75842713, 16.90959037, 16.77740237,
       14.8441752 , 17.83177668, 16.79637732, 17.84813616, 21.06375561,
       17.94214009, 17.08030864, 15.46657595, 21.38986221, 15.82584278,
       15.85687487, 14.43494791, 17.1908025 , 13.47761693, 17.00242386,
       16.622159  , 17.36167363, 18.59974719, 18.80350917, 18.96573893,
       19.85378244, 15.36407422, 16.07197145, 20.73065969, 18.02

Поскольку мы проверяем гипотезу относительно математического ожидания нормально распределённой случайной величины с известной дисперсией, будем использовать $Z$ статистику:

$$Z = \dfrac{\overline{X} - \mu}{\sigma / \sqrt{n}},$$

При условии верности нулевой гипотезы статистика имеет стандартное нормальное распределение.

Рассчитаем статистику:

In [23]:
def statistic(samples):
    return (samples.mean() - 17) / (2 / np.sqrt(n))

In [24]:
z = statistic(samples)
z

3.108411727714806

Поскольку используется правосторонняя критическая область $\Omega_\alpha = (t_{1 - \alpha}, \infty)$ и $\alpha = 0.025$ :

In [30]:
alpha = 0.025

T = stats.norm.ppf(1 - alpha)

f"Критическая область ({T}, '+inf')"

"Критическая область (1.959963984540054, '+inf')"

Значение статистики $\ Z$ попало в критическую область (3.1 > 1.96), поэтому гипотезу, что шарики для подшипников, изготовленные автоматическим станком, имеют средний диаметр 17 мм, отвергаем.